### DISCUSSION TITLE
Credential Page Two Time Coming

### DISCUSSION THREAD

**Discussion by arunrajac** [zitadel/zitadel#8289] [Q&A]
Could you please help me in my website integrated zitadel after logged again its redirecting to login page.

**Comment by hifabienne** [comment]
Hei @arunrajac 
Can you please share some more details?
What do you try to achieve, how did you integrate?
And what happens exactly?

**Reply by arunrajac** [reply]
sample angular github code i used.
after logged in it redirecting to my website but im not able to get true in this method "oauth.hasValidAccessToken()"
always im getting false and its redirecting to login oage

**Reply by hifabienne** [reply]
@peintnermax can you help here? which information do we need to help?

**Reply by fforootd** [reply]
@arunrajac please share us a screenshot of the client config in zitadel.

Also a HAR file could help understanding this.

**Reply by arunrajac** [reply]
import {
    APP_INITIALIZER,
    EnvironmentProviders,
    importProvidersFrom,
    inject,
    Provider,
} from '@angular/core';
import { AuthConfig, OAuthModule, OAuthStorage } from 'angular-oauth2-oidc';
import {
    StatehandlerProcessorService,
    StatehandlerProcessorServiceImpl,
} from 'app/services/statehandler-processor.service';
import {
    StatehandlerService,
    StatehandlerServiceImpl,
} from 'app/services/statehandler.service';
import { StorageService } from 'app/services/storage.service';
import {environment} from "../../../environments/environment";

const authProdConfig: AuthConfig = {
    scope: 'openid profile email offline_access urn:zitadel:iam:user:resourceowner',
    responseType: 'code',
    oidc: true,
    clientId: '',
    issuer: environment.authURL,
    redirectUri: `https://${window.location.hostname}/overview`,
    postLogoutRedirectUri: `https://${window.location.hostname}/sign-in`,
    requireHttps: true,
    sessionChecksEnabled: true,
};
const stateHandlerFn = (stateHandler: StatehandlerService) => {
    return () => {
        return stateHandler.initStateHandler();
    };
};

export const provideZitadel = (): Array<Provider | EnvironmentProviders> => {
    return [
        importProvidersFrom(
            OAuthModule.forRoot({
                resourceServer: {
                    allowedUrls: [
                        `${window.location.hostname}/admin/v1`,
                        `${window.location.hostname}/management/v1`,
                        `${window.location.hostname}/auth/v1/`,
                    ],
                    sendAccessToken: true,
                },
            })
        ),
        {
            provide: AuthConfig,
            useValue: authProdConfig,
        },
        {
            provide: StatehandlerProcessorService,
            useClass: StatehandlerProcessorServiceImpl,
        },
        {
            provide: StatehandlerService,
            useClass: StatehandlerServiceImpl,
        },
        {
            provide: OAuthStorage,
            useClass: StorageService,
        },
        {
            provide: APP_INITIALIZER,
            useFactory: stateHandlerFn,
            multi: true,
            deps: [StatehandlerService],
        },
    ];
};

@fforootd can you please help me here

**Reply by fforootd** [reply]
Can you please share a screenshot of how you configured the client in zitadels UI?

**Reply by fforootd** [reply]
Thanks, the first thing that strikes me is that the type might need to be "native".

Can you share what response you get from zitadel when you are redirected back to your app?

Can you share me a har file or a screenshot of your browsers network tab?

**Reply by arunrajac** [reply]
@fforootd sir can we connect teams if possible. if possible can you please share your free time sir.

**Reply by fforootd** [reply]
Hey, please provide us with the requested data first, we can see from there how it goes.

You can send privately to florian@zitadel.com

**Reply by fforootd** [reply]
Hey, I checked the file and to me it looks like your app "just restarts" the authorization process again. Even though it gets the code from zitadel.

Do you see logs in your app on what happens? I guess it must be an error in stdout/console logwhen this happens.

**Reply by fforootd** [reply]
I just noted... you app sends the following callback url `https://{yourdomain}/overview` I guess that should be more like `https://{yourdomain}/auth/callback`

This also is in your code snippet above `redirectUri: https://${window.location.hostname}/overview,`

In our example we use `redirectUri: 'http://localhost:4200/auth/callback',`

**Reply by arunrajac** [reply]
As per your feedback we have changed the redirect url to /auth/callback but still same issue.

We have tried with the zitadel angular lib and it working, but as per our project requirement we have parent and child router.
If we add child router with zitadel angular lib then same issue happening.
Please find below router which using for our project.
@fforootd

**Reply by arunrajac** [reply]
@fforootd sir any example code for angular 17 standalone component?

**Reply by fforootd** [reply]
Maybe @peintnermax has a clue about what you are asking, I lack the answer for your case of the top of my mind.

Though its clear that this is an implementation issue and not zitadel not behaving as intended 😁 

Bt.w please stop tagging me with each post, that does not help in any way or form

**Reply by peintnermax** [reply]
One thing that comes to my mind is that your angular app may not completing the code flow due to a missconfigured router.
Can you make sure that the callback (in your case /auth/callback) is a guarded page and from there the authenticate() is called again. This way your application can exchange the code with a token and complete the flow. 
https://zitadel.com/docs/examples/login/angular#add-login-to-your-application explains this in the section `Add login to your application`

**Reply by peintnermax** [reply]
addition:

```
{
        path: 'user',
        component: UserComponent,
        canActivate: [AuthGuard],
    },
    { 
        path: 'auth/callback',
        redirectTo: 'user'
    },
```

note that the callback redirects to the user page which is guarded and continues to check the auth state.
